File: go/reddit.go
package main

import (
	"fmt"
	"log"
	"runtime"
	"sync"

	"github.com/PuerkitoBio/goquery"
	"github.com/fatih/color"
)

var redditUrl string = "http://reddit.com/r/programming"

type Article struct {
	Title       string
	Link        string
	Description string
}

func NewArticle(title string, link string) Article {
	a := Article{
		Title: title,
		Link:  link,
	}

	a.getDescription()

	return a
}

func (a *Article) getDescription() {
	doc, _ := goquery.NewDocument(a.Link)

	metas := doc.Find("meta[name=description]")

	if len(metas.Nodes) > 0 {
		a.Description, _ = metas.Attr("content")
	}
}

func (a Article) Show() {
	green := color.New(color.FgGreen).SprintFunc()
	blue := color.New(color.FgBlue).SprintFunc()
	fmt.Printf("* [%s] (%s)", green(a.Title), blue(a.Link))
	fmt.Println(a.Description + "\n")
}

func main() {

	numCPU := runtime.NumCPU()
	runtime.GOMAXPROCS(numCPU)

	doc, err := goquery.NewDocument(redditUrl)
	if err != nil {
		log.Fatal(err)
	}

	var wg sync.WaitGroup
	links := doc.Find("a.title")
	wg.Add(len(links.Nodes))

	articlesChan := make(chan Article, len(links.Nodes))

	for i := range links.Nodes {
		s := links.Eq(i)
		go func(wg *sync.WaitGroup, arts chan Article, node *goquery.Selection) {
			defer wg.Done()

			title := s.Text()
			link, _ := s.Attr("href")

			a := NewArticle(title, link)

			fmt.Printf("fetching %s\n", link)

			arts <- a
		}(&wg, articlesChan, s)
	}
	wg.Wait()
	close(articlesChan)

	for article := range articlesChan {
		article.Show()
	}
}

File: reddit.py
import urllib
from BeautifulSoup import BeautifulSoup as Soup
from beautifulsoupselect import select
from termcolor import colored


class Article():
    title = None
    link = None
    description = "No description"

    def __init__(self, title, link):
        self.title = title
        self.link = link
        self.get_meta_description()

    def get_meta_description(self):
        soup = Soup(urllib.urlopen(self.link))
        metas = select(soup, 'meta[name=description]')

        if len(metas) > 0:
            self.description = metas[0].get('content')

        print "\r fetching " + self.link


def main():
    hacker_news_url = "https://www.reddit.com/r/programming"
    soup = Soup(urllib.urlopen(hacker_news_url))
    links = select(soup, 'a.title')
    articles = []

    for link in links:
        title = link.string
        href = link.get('href')
        article = Article(title, href)
        articles.append(article)

    for article in articles:
        print " * [" + colored(article.title, 'green') + "] (" + colored(article.link, 'blue') + ")"
        print article.description


if __name__ == '__main__':
    main()
